libpromises/evalfunction.c

```diff
@@ -6659,6 +6658,72 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
     return (FnCallResult) { FNCALL_SUCCESS, (Rval) { json, RVAL_TYPE_CONTAINER } };
 }
 
+static FnCallResult FnCallClassExprFilter(EvalContext *ctx,
+                                          ARG_UNUSED Policy const *policy,
+                                          ARG_UNUSED FnCall const *fp,
+                                          Rlist const *args)
+{
+    char const *path = RlistScalarValue(args);
+    size_t const col = IntFromString(RlistScalarValue(args->next));
+    char const *delim = RlistScalarValue(args->next->next);
+    bool has_heading = BooleanFromString(RlistScalarValue(args->next->next->next));
+    int const sort_col = IntFromString(RlistScalarValue(args->next->next->next->next));
+
+    Seq *heading_seq = NULL;
+    char *line;
+    JsonElement *json = JsonArrayCreate(50);
+
```

Comment:
JsonArrayCreate() calls xcalloc for each element, so need to free before returning on line 6679 below or any other return paths below.

Reply:
I will focus on logic for now, but you probably want to double-check memory alloc/free such as JsonArrayCreate() and GetCsvLineNext() and maybe others.